src/Extensions/Microsoft.Diagnostics.Monitoring.Extension.Common/EgressHelper.cs

```diff
@@ -35,8 +38,28 @@ private static async Task<int> Egress<TOptions>(EgressProvider<TOptions> provide
             EgressArtifactResult result = new();
             try
             {
-                string jsonConfig = Console.ReadLine();
-                ExtensionEgressPayload configPayload = JsonSerializer.Deserialize<ExtensionEgressPayload>(jsonConfig);
+                StdInStream = Console.OpenStandardInput();
+
+                int dotnetMonitorPayloadProtocolVersion;
+                long payloadLengthBuffer;
+                byte[] payloadBuffer;
+
+                using (var reader = new BinaryReader(StdInStream, Encoding.UTF8, true))
+                {
+                    dotnetMonitorPayloadProtocolVersion = reader.ReadInt32();
+                    payloadLengthBuffer = reader.ReadInt64();
+
+                    payloadBuffer = new byte[payloadLengthBuffer];
+
+                    reader.Read(payloadBuffer, 0, (int)payloadLengthBuffer); // this should be safe, given we know the payload is small
```

Comment:
Let's not use the `BinaryReader` to read out the payload since we aren't using the `BinaryWriter` to write the payload. Basically, the same methodology that we use to write is the same methodology that we should use to read.

Reply:
I re-integrated `ReadExactlyAsync` for this - given that we still need it, what's the advantage of using Binary*er? My initial thought was it would enable us to get away from the manual `ReadExactlyAsync` method, but it seems like it's only solving part of the problem.

Reply:
`BinaryWriter`/`BinaryReader` are good for writing protocols for exact serialization due to endianness concerns. `BinaryWriter`/`BinaryReader` will always write/read using little endian. Thus we can say "the first 4 bytes are the protocol version and its in little endian form".

I think that Windows is always little endian whereas Linux and Mac can be either depending on the processor its running on. While this is probably okay for .NET code to not care about it (because the .NET runtime will do the same thing on both sides of the communication so long as we use the same serialization methodology), native code extensions will not be as tolerant.